### ISSUE TITLE
How to create a record that is supposed to contain other records and lists ?

### ISSUE THREAD

**Issue by ksachdeva** [rangle/typed-immutable-record#19]
interface IChild {
   someInt: number;
   someStr: string;
}

interface IParent {
  child: IChild;
  children: IChild[];
}

// to recordify

export interface IParentRecord extends TypedRecord<IParentRecord>, IParent {
}

const anImmutableRecord =  recordify<IParent, IParentRecord>({
  child: ? ...
  children: ?...
});

How to create anImmutableRecord that would contain other immutable records and imutable list ?

Regards & thanks
Kapil

**Comment by danielfigueiredo** [comment]
Hello, considering the current Record API provided by (immutable.js)[https://facebook.github.io/immutable-js/docs/#/Record] 
it doesn't provide a nested way of building records.
I have this (issue)[https://github.com/rangle/typed-immutable-record/issues/17] open to deal with nested trees, but not automatic solution so far.
Calling fromJS will make Immutable always create Map / Lists
At this point if you want to create a immutable List, you would need to manually create the list and assign the values, the interface would also need to accept the type, which can be done either with union operator in typescript or just use the immutable List

```typescript

interface IChild {
someStr: string;
}

interface IParent {
child: IChild;
children: List<IChild> | IChild[];
}

// to recordify
export interface IChildRecord extends TypedRecord<IChildRecord>, IChild {}
export interface IParentRecord extends TypedRecord<IPerentRecord>, IParent {}

const myChildRecordFactory = makeTypedFactory<IChild, IChildRecord>({someStr: null});
const myParentRecord = recordify<IParent, IPerentRecord>({
  child: myChildRecordFactory({someStr: 'My Child'}),
  children: List([
    myChildRecordFactory({someStr: 'My Child 2'}),
    myChildRecordFactory({someStr: 'My Child 3'}),
    myChildRecordFactory({someStr: 'My Child 4'})
});
```

So basically you need to create the structure yourself, maybe composing this creation will make the code better organized and reusable (different then my solution above)
:)

**Comment by hirtie-maxim** [comment]
@ksachdeva, as a workaround can be used something like this
`import { List, Iterable } from 'immutable';

export interface TypedList<T extends TypedList<T>> extends List<T> {
    set(index: number, value: T): List<T>;
    delete(index: number): List<T>;
    remove(index: number): List<T>;
    insert(index: number, value: T): List<T>;
    clear(): List<T>;
    push(...values: T[]): List<T>;
    pop(): List<T>;
    unshift(...values: T[]): List<T>;
    shift(): List<T>;
    update(updater: (value: List<T>) => List<T>): List<T>;
    update(index: number, updater: (value: T) => T): List<T>;
    update(index: number, notSetValue: T, updater: (value: T) => T): List<T>;
    merge(...iterables: Iterable.Indexed<T>[]): List<T>;
    merge(...iterables: Array<T>[]): List<T>;
    mergeWith(
      merger: (previous?: T, next?: T, key?: number) => T,
      ...iterables: Iterable.Indexed<T>[]
    ): List<T>;
    mergeWith(
      merger: (previous?: T, next?: T, key?: number) => T,
      ...iterables: Array<T>[]
    ): List<T>;
    mergeDeep(...iterables: Iterable.Indexed<T>[]): List<T>;
    mergeDeep(...iterables: Array<T>[]): List<T>;
    mergeDeepWith(
      merger: (previous?: T, next?: T, key?: number) => T,
      ...iterables: Iterable.Indexed<T>[]
    ): List<T>;
    mergeDeepWith(
      merger: (previous?: T, next?: T, key?: number) => T,
      ...iterables: Array<T>[]
    ): List<T>;
    setSize(size: number): List<T>;
    setIn(keyPath: Array<any>, value: any): List<T>;
    setIn(keyPath: Iterable<any, any>, value: any): List<T>;
    deleteIn(keyPath: Array<any>): List<T>;
    deleteIn(keyPath: Iterable<any, any>): List<T>;
    removeIn(keyPath: Array<any>): List<T>;
    removeIn(keyPath: Iterable<any, any>): List<T>;
    updateIn(
      keyPath: Array<any>,
      updater: (value: any) => any
    ): List<T>;
    updateIn(
      keyPath: Array<any>,
      notSetValue: any,
      updater: (value: any) => any
    ): List<T>;
    updateIn(
      keyPath: Iterable<any, any>,
      updater: (value: any) => any
    ): List<T>;
    updateIn(
      keyPath: Iterable<any, any>,
      notSetValue: any,
      updater: (value: any) => any
    ): List<T>;
    mergeIn(
      keyPath: Iterable<any, any>,
      ...iterables: Iterable.Indexed<T>[]
    ): List<T>;
    mergeIn(
      keyPath: Array<any>,
      ...iterables: Iterable.Indexed<T>[]
    ): List<T>;
    mergeIn(
      keyPath: Array<any>,
      ...iterables: Array<T>[]
    ): List<T>;
    mergeDeepIn(
      keyPath: Iterable<any, any>,
      ...iterables: Iterable.Indexed<T>[]
    ): List<T>;
    mergeDeepIn(
      keyPath: Array<any>,
      ...iterables: Iterable.Indexed<T>[]
    ): List<T>;
    mergeDeepIn(
      keyPath: Array<any>,
      ...iterables: Array<T>[]
    ): List<T>;
    withMutations(mutator: (mutable: List<T>) => any): List<T>;
    asMutable(): List<T>;
    asImmutable(): List<T>;
};

export function makeTypedFactoryList<E, T extends TypedList<T> & E>(obj: E, name?: string): (val?: E) => TypedList<T> {
  const immutableList = List();
  return function TypedFactory(val: E = null): TypedList<T> {
    return immutableList.push(val) as TypedList<T>;
  };
};`

**Comment by Kevin-Deason** [comment]
How would one use the above makeTypedFactoryList?

I'm having quite some difficulty creating a similar scenario as ksachdeva

**Comment by mikesmo** [comment]
I used the following strategy:

` 
interface ICreditCard {
    name: string;
    number: string;
}

interface ICustomer {
    name: string;
    creditCards: List<ICreditCard> | ICreditCard[];
}

interface ICreditCardRecord extends TypedRecord<ICreditCardRecord>, ICreditCard {}
interface ICustomerRecord extends TypedRecord<ICustomerRecord>, ICustomer {}

function toTypedList<T>(array : List<T> | T[], factory) : List<T> {
  return List<T>((array as T[]).map(a => factory(a)));
}
 
function CustomerFactory(customer: ICustomer): ICustomerRecord {
    return recordify<ICustomer, ICustomerRecord>({
        name: customer.name,
        creditCards: toTypedList(customer.creditCards, (creditCard: ICreditCard) =>
            recordify<ICreditCard, ICreditCardRecord>(creditCard))
    });
}
`

then:

`
let customer = {
            name: 'John Smith',
            creditCards: [{
                name: 'J P Smith',
                number: '12345667112435'
            }]
        };

        let customerRecord = CustomerFactory(customer);

        let creditCardRecords = customerRecord.creditCards as List<ICreditCard>;
        creditCardRecords.forEach(card => {
            console.log(card.name);
            console.log(card.number);
        });
`

The issues with this are:

1. There is a lot of typing that needs to be done once an object becomes deeply nested.
2. You have to cast creditCards to List<ICreditCard> in order to access the immutable operations.

**Comment by danielfigueiredo** [comment]
@mikesmo that is correct, you just need to build the object structure
After few issues get resolved in typescript we can have a much better API compared to what we can do at this point

**Comment by danielfigueiredo** [comment]
Maybe this PR https://github.com/rangle/typed-immutable-record/pull/25 might help to solve this issue.

**Comment by danielfigueiredo** [comment]
@knight-bubble if you wanna make a PR for the TypedList, that would be pretty cool
